swap/protocol.go

```diff
@@ -0,0 +1,167 @@
+
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	contract "github.com/ethersphere/swarm/contracts/swap"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// ErrEmptyAddressInSignature is used when the empty address is used for the chequebook in the handshake
+var ErrEmptyAddressInSignature = errors.New("empty address in handshake")
+
+// ErrInvalidHandshakeMsg is used when the message received during handshake does not conform to the
+// structure of the HandshakeMsg
+var ErrInvalidHandshakeMsg = errors.New("invalid handshake message")
+
+// Spec is the swap protocol specification
+var Spec = &protocols.Spec{
+	Name:       "swap",
+	Version:    1,
+	MaxMsgSize: 10 * 1024 * 1024,
+	Messages: []interface{}{
+		HandshakeMsg{},
+		EmitChequeMsg{},
+	},
+}
+
+// Protocols is a node.Service interface method
+func (s *Swap) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{
+		{
+			Name:    Spec.Name,
+			Version: Spec.Version,
+			Length:  Spec.Length(),
+			Run:     s.run,
+		},
+	}
+}
+
+// APIs is a node.Service interface method
+func (s *Swap) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "swap",
+			Version:   "1.0",
+			Service:   NewAPI(s),
+			Public:    false,
+		},
+	}
+}
+
+// Start is a node.Service interface method
+func (s *Swap) Start(server *p2p.Server) error {
+	log.Info("Swap service started")
+	return nil
+}
+
+// Stop is a node.Service interface method
+func (s *Swap) Stop() error {
+	return nil
+}
+
+// verifyHandshake verifies the chequebook address transmitted in the swap handshake
+func (s *Swap) verifyHandshake(msg interface{}) error {
+	handshake, ok := msg.(*HandshakeMsg)
+	if !ok {
+		return ErrInvalidHandshakeMsg
+	}
+
+	if (handshake.ContractAddress == common.Address{}) {
+		return ErrEmptyAddressInSignature
+	}
+
+	return s.verifyContract(context.TODO(), handshake.ContractAddress)
+}
+
+// run is the actual swap protocol run method
+func (s *Swap) run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	protoPeer := protocols.NewPeer(p, rw, Spec)
+
+	handshake, err := protoPeer.Handshake(context.TODO(), &HandshakeMsg{
+		ContractAddress: s.owner.Contract,
+	}, s.verifyHandshake)
+	if err != nil {
+		return err
+	}
+
+	response, ok := handshake.(*HandshakeMsg)
+	if !ok {
+		return ErrInvalidHandshakeMsg
+	}
+
+	beneficiary, err := s.getContractOwner(context.TODO(), response.ContractAddress)
+	if err != nil {
+		return err
+	}
+
+	swapPeer := NewPeer(protoPeer, s, s.backend, beneficiary, response.ContractAddress)
+	s.addPeer(swapPeer)
+	defer s.removePeer(swapPeer)
+
+	return swapPeer.Run(swapPeer.handleMsg)
+}
+
+func (s *Swap) removePeer(p *Peer) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	delete(s.peers, p.ID())
+}
+
+func (s *Swap) addPeer(p *Peer) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.peers[p.ID()] = p
+}
+
+func (s *Swap) getPeer(id enode.ID) (*Peer, error) {
+	var err error
+	peer := s.peers[id]
```

Comment:
then:
```go
peer, ok := s.peers[id]
return peer, ok
```
or
```go
peer, ok := s.peers[id]
if !ok {
 return nil, ok
}
return peer, ok
```

Reply:
also, this `getPeer` is not under lock! make sure you `s.lock.RLock()` then `RUnlock()`

Reply:
first part is done. We had an RLock here in the past but the issue was that this function is called by the `Add` function which already has a (write) lock on Peer thus causing a deadlock situation.

Reply:
i have some objections about the way we are using locks. i'd like to review this at some not too distant point in the future.

Reply:
created issue #1663 to deal with this.

Reply:
@ralph-pichler a deadlock situation means you are misusing locks, not that a call should not be under lock.
if you are having a hard time to formulate a correct usage of locks, the solution then should not be "not to use locks." in this case we do not use locks because they are cool to use but because maps in go are not thread-safe. under certain conditions retrieving an item from the map might end up returning nothing or panicing. you can read more [here](https://stackoverflow.com/questions/36167200/how-safe-are-golang-maps-for-concurrent-read-write-operations).

From what I see the `Add` method should not be under lock, at least not all of it, also, you might want to consider adding another mutex to just synchronize access to the `peers` map and use that mutex only in `addPeer`, `removePeer`, `getPeer`.

@mortelli please deal with the problem now. I'm not OK with deferring this to a later PR.